ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+                (segment.z_right, entry.z_right),
+            ] {
+                // unwrap safe since index are guaranteed to be within bounds
+                let values_in_array = array_value.get_mut(position_bit_array..length).unwrap();
+                values_in_array.clone_from_bitslice(segment_value.0);
+            }
+        } else {
+            // segments are multiples of 256
+            let length_in_entries = length >> 8;
+            for i in 0..length_in_entries {
+                let entry_option = &mut self.vec[position_vec + i];
+                // make sure we don't overwrite values
+                debug_assert!(entry_option.is_none());
+                *entry_option = Some(UnverifiedValues::new_from_bitslices(
+                    &segment.x_left.0[256 * i..256 * (i + 1)],
+                    &segment.x_right.0[256 * i..256 * (i + 1)],
+                    &segment.y_left.0[256 * i..256 * (i + 1)],
+                    &segment.y_right.0[256 * i..256 * (i + 1)],
+                    &segment.prss_left.0[256 * i..256 * (i + 1)],
+                    &segment.prss_right.0[256 * i..256 * (i + 1)],
+                    &segment.z_right.0[256 * i..256 * (i + 1)],
+                ));
+            }
+        }
+    }
+
+    /// `get_unverified_field_value` converts a `UnverifiedValuesStore` into an iterator over `UnverifiedFieldValues`
+    /// compatible with DZKPs
+    fn get_unverified_field_values<DF: DZKPBaseField>(
+        &self,
+    ) -> impl Iterator<Item = DF::UnverifiedFieldValues> + '_ {
+        self.vec
+            .iter()
+            .filter_map(|x| x.as_ref())
+            .map(UnverifiedValues::convert::<DF>)
+    }
+}
+
+/// `Batch` collects a batch of `UnverifiedValuesStore` in a hashmap.
+/// The size of the batch is limited due to the memory costs and verifier specific constraints.
+///
+/// Corresponds to `AccumulatorState` of the MAC based malicious validator.
+#[derive(Clone, Debug)]
+struct Batch {
+    chunk_size: usize,
+    inner: HashMap<Gate, UnverifiedValuesStore>,
+}
+
+#[allow(dead_code)]
+impl Batch {
+    fn new(chunk_size: usize) -> Self {
+        Self {
+            chunk_size,
+            inner: HashMap::<Gate, UnverifiedValuesStore>::new(),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+            || self
+                .inner
+                .iter()
+                .fold(true, |acc, (_, value)| acc && value.is_empty())
+    }
+
+    fn push(&mut self, gate: &Gate, record_id: RecordId, segment: Segment) {
+        // get value_store & create new when necessary
+        if self.inner.get(gate).is_none() {
+            let value_store = UnverifiedValuesStore::new(self.chunk_size);
+            self.inner.insert(gate.clone(), value_store);
+        }
+
+        // insert segment
+        self.inner
+            .get_mut(gate)
+            .unwrap()
+            .insert_segment(record_id, segment);
```

Comment:
You can use `self.inner.entry(gate).or_insert_with(/* closure */)` to avoid the `is_none` / `unwrap` combo.